### PR TITLE
feat: refactor to ts, test against wider range of `@typescript-eslint`

### PR DESCRIPTION
**Pull request by antfu**
As https://github.com/sweepline/eslint-plugin-unused-imports/issues/82#issuecomment-2243091748

I am sorry for a rather big PR, but changes are interconnected and I thought it would be better to do them all together. Happy to explain if you have any doubts. These changes allow us to have more confidence when introducing changes, and help with long term maintainability.

This PR changes:

- Refactor the codebase to TypeScript and ESM
- Publish both CJS and ESM
- Migrate from `jest` to `vitest` (as it has better TS and ESM support)
- Run a test matrix to also test against old versions of `@typescript-eslint`
- Relax the version range of `@typescript-eslint` (as logic-wise there is no requirements on the version)
- Switch to `pnpm` (better handling for package alias)
- No longer deps on `eslint-rule-composer`, as it seems to be unmaintained, and we don't use a lot from it. It's simpler to direct do that in-house.
- Removes `engines` field in `package.json`. As our code is rather pure and `eslint` itself would enforce the engines requirements already

### CODE REVIEW

**Comment by sweepline** [comment]
This is really nice, thanks!
I'm wondering if we could get away with just a minor version as its not breaking anything for version 4.x.x users?

**Review by sweepline — APPROVED** [review]

**Comment by antfu** [comment]
Totally, I think it would definitely be ok to be a minor version. I think it's more like how we communicate to the users.

File: README.md [hunk 1]
```diff
@@ -4,6 +4,7 @@ Find and remove unused es6 module imports. It works by splitting up the `no-unus
 
 ## _Versions_
 
+-   Version 5.x.x is for eslint 9 with @typescript-eslint/eslint-plugin v8 to v5
 -   Version 4.x.x is for eslint 9 with @typescript-eslint/eslint-plugin 8
```
Comment (antfu) [inline README.md:8]:
```suggestion
-   Version 4.1.x is for eslint 9 with @typescript-eslint/eslint-plugin v8 to v5
```

We could do this as well

Comment (sweepline) [reply README.md:8]: Yeah that is good.

**Comment by antfu** [comment]
@sweepline I think it should be good to go? Should I handle the merge and release? If so, can I have permission to the npm package (my id is `antfu`). Thanks!

**Comment by sweepline** [comment]
@antfu Yeah it looks good. I invited you on npm.

**Comment by clemens** [comment]
Observation: In its current state, the README seems to be missing the information on what version 4.0.x is intended for. 3.x.x and 4.1.x are covered, but not 4.0.x.

**Pull request merged**